Text/PrettyPrint/Compact/Core.hs

```diff
@@ -1,83 +1,117 @@
-{-# LANGUAGE ScopedTypeVariables, TypeSynonymInstances, FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, ViewPatterns #-}
-module Text.PrettyPrint.Compact.Core(Layout(..),Document(..),Doc) where
+{-# LANGUAGE ScopedTypeVariables, TypeSynonymInstances, FlexibleContexts, FlexibleInstances, GeneralizedNewtypeDeriving, ViewPatterns, DeriveFunctor, DeriveFoldable, DeriveTraversable #-}
+module Text.PrettyPrint.Compact.Core(Layout(..),Render(..),Document(..),Doc,annotate) where
 
-import Data.List (sort,groupBy,intercalate,minimumBy)
+import Data.List (sortOn,groupBy,minimumBy)
 import Data.Function (on)
 import Data.Semigroup
 import Data.Sequence (singleton, Seq, viewl, viewr, ViewL(..), ViewR(..), (|>))
 import Data.String
 import Data.Foldable (toList)
 
-newtype L = L (Seq String) -- non-empty sequence
-  deriving (Eq,Ord,Show)
+-- | Annotated string. We keep annotated segments in a separate list.
+--
+-- First argument is length.
+--
+data AS a = AS !Int [(Maybe a, String)]
```

Comment:
This data type needs to be explained and/or simplified:
- what are the invariants?
- why do we have a list? 
- why do you have a Maybe instead of using the mempty of a?